### DISCUSSION TITLE
Y4M encoding and transcoding from arbitrary formats with VVenC

### DISCUSSION THREAD

**Discussion by adamjw24** [fraunhoferhhi/vvenc#116] [Show and tell]
VVenC does not support the Y4M format. Support is not planned, for a variety of reasons, including lack of official specification of the Y4M format.

For people needing Y4M encoding support, please use a support script, e.g. like the one provided below.

The call semantics for that specific script would be: `encode_y4m.sh input.y4m [vvenc options]`, where `vvenc options` represent the intended vvenc encoding options (not related to the input description, which will be provided by the script), so e.g. `--preset medium -f 33 -o out.266` to encode first 33 frames using the medium preset and save the bitstream as `out.266`.

Please mind that the script requires VVenC 1.3 (to be released soon), or at least 5a21fb6. For earlier versions the fractional semantics of the frame-rate (e.g. 60/1 or 60000/1001) have to be converted into an integer frame-rate.

```sh
#!/bin/sh

##### script configuration

VVENC=../vvenc/install/bin/vvencapp

##### parse parameters

INPUT=$1
shift
VVENC_OPTS=$*

##### get input info

eval $(ffprobe -v error -select_streams v:0 -show_entries stream=width,height,r_frame_rate -of default=nw=1 $INPUT)
echo "Input (${INPUT}): ${width}x${height} at ${r_frame_rate}"
echo

##### encoding

ffmpeg -hide_banner -loglevel error -i ${INPUT} -f rawvideo -pix_fmt yuv420p10le - | \
$VVENC -c yuv420_10 -s ${width}x${height} --fps ${r_frame_rate} -i - ${VVENC_OPTS}

```

**Comment by FranceBB** [comment]
I know this is a slightly different project, but one of the reasons why people would like to have y4m is to pipe and not having to specify every single option every time with the risk of typos etc. On this regard, one of the things that would also be hugely, I mean hugely appreciated by the whole Doom9 community and people around the world is Avisynth support, namely being able to use Avisynth as a frameserver to encode with VVEnc. If we take a look at x262, x264, x265 they all support Avisynth. Now, we know that x266 ain't out yet, therefore one real advantage VVEnc could have would be supporting Avisynth and/or use LAV as inputs, at least on Windows.

p.s thank you for making this free and open source. ;)

**Reply by adamjw24** [reply]
Now, correct me if I'm wrong, but the above script would work for .avs files too, no? Actually it would support transcoding of any format that provides the `width`, `height`, and `r_frame_rate` properties into VVC.

Its the first time I looked up Avisynth, so not sure how it works. From the quick overview, I understand that ffmpeg can just read .avs files like normal video files, right? Please excuse me if my understanding of Avisynth is too bad and this doesn't really make sense.

If I have a while I might try and create a .bat version of the above script.

**Comment by FranceBB** [comment]
That is correct. FFMpeg definitely can use -i "AVS Script.avs" and read Avisynth Scripts as if they were a normal video 'cause the frameserver would just be serving frames to FFMpeg and of course FFMpeg could pipe them again to VVEnc, however some encoders like x26x can open them directly without going through FFMpeg.
Nonetheless, for the time being, a BAT would be appreciated, definitely, so thanks for looking into this. :)

**Reply by adamjw24** [reply]
This produces some broken pipe errors on the ffmpeg side if encoding less than the whole input, but otherwise works

```batch
@echo off

REM script configuration

set VVENC=vvencapp.exe
set FFMPEG=ffmpeg.exe
set FFPROB=ffprobe.exe

REM parse parameters

set INPUT=%1
for /f "tokens=1,* delims= " %%a in ("%*") do set VVENC_OPTS=%%b

REM get input info

for /f %%i in ('%FFPROB% -v error -select_streams v:0 -show_entries stream^=width^,height^,r_frame_rate -of default^=nw^=1 %INPUT%') do set %%i

echo Input (%INPUT%): %width%x%height% at %r_frame_rate%

REM encoding

%FFMPEG% -hide_banner -loglevel error -i %INPUT% -f rawvideo -pix_fmt yuv420p10le - | ^
%VVENC% -c yuv420_10 -s %width%x%height% --fps %r_frame_rate% -i - %VVENC_OPTS%
```

**Reply by georgysaumen** [reply]
> 

Hello, any idea with this ?

![Untitled](https://user-images.githubusercontent.com/86090065/158155843-6c8f68e8-2e9c-4624-9fcf-a0c339fa5ff9.png)

**Reply by adamjw24** [reply]
How did you call the script? The error message it pretty self-explanatory. You need to specify the output.

**Reply by georgysaumen** [reply]
> How did you call the script? The error message it pretty self-explanatory. You need to specify the output.

I just copy your bat file and run

@echo off

REM script configuration

set VVENC=vvencapp.exe
set FFMPEG=ffmpeg.exe
set FFPROB=ffprobe.exe

REM parse parameters

set INPUT=111video.mp4
for /f "tokens=1,* delims= " %%a in ("%*") do set VVENC_OPTS=%%b

REM get input info

for /f %%i in ('%FFPROB% -v error -select_streams v:0 -show_entries stream^=width^,height^,r_frame_rate -of default^=nw^=1 %INPUT%') do set %%i

echo Input (%INPUT%): %width%x%height% at %r_frame_rate%

REM encoding

%FFMPEG% -hide_banner -loglevel error -i %INPUT% -f rawvideo -pix_fmt yuv420p10le - | ^
%VVENC% -c yuv420_10 -s %width%x%height% --fps %r_frame_rate% -i - %VVENC_OPTS%

**Reply by georgysaumen** [reply]
> How did you call the script? The error message it pretty self-explanatory. You need to specify the output.

Oh, it run, this only create raw file ?

**Reply by adamjw24** [reply]
Like please at least read how to use the script, as explained above.

You download the script, save it as a bat file (under `name.bat`), and call the script instead of the encoder (`name.bat input.mp4 [vvenc options]`, no need to specify size, framerate, or input bitdepth).

**Reply by georgysaumen** [reply]
> Like please at least read how to use the script, as explained above.
> 
> You download the script, save it as a bat file (under `name.bat`), and call the script instead of the encoder (`name.bat input.mp4 [vvenc options]`, no need to specify size, framerate, or input bitdepth).

I need to wait until finish and use that video to watch ?

![Untitled](https://user-images.githubusercontent.com/86090065/158158008-caf964c8-3716-438a-99c3-13730cd8fb1c.png)

**Reply by adamjw24** [reply]
Yes. Except there is no real way to watch VVC yet.

**Reply by georgysaumen** [reply]
> Yes. Except there is no real way to watch VVC yet.

So how can i watch now ? :))) And can I ask how long do you encode your video, for example your video is 1 hour 30 minutes to 2 hours? On my video, just 2 minutes but I see they encode so slow, 2 hours but just 4mb and dont finish yet. I use I7-9750h

**Reply by adamjw24** [reply]
> So how can i watch now ? :))) 

Well, we have the decoder for that, but its useless by itself (or, lets say, of academical value), you'd need it integrated in a player. You should ask the player and framework maintainers why VVC is not supported. You could try http://dx.doi.org/10.1145/3474085.3478320 for playback, but its an expert solution.

> And can I ask how long do you encode your video, for example your video is 1 hour 30 minutes to 2 hours? On my video, just 2 minutes but I see they encode so slow, 2 hours but just 4mb and dont finish yet. I use I7-9750h

About the speed. Which preset do you use? faster runs at around 5fps, fast around 1-2 and medium around 1/2 fps. Depending on your CPU and number of used threads.

**Reply by georgysaumen** [reply]
> > So how can i watch now ? :)))
> 
> Well, we have the decoder for that, but by its useless by itself (or, lets say, of academical value), you'd need it integrated in a player. You should ask the player and framework maintainers why VVC is not supported. You could try http://dx.doi.org/10.1145/3474085.3478320 for playback, but its an expert solution.
> 
> > And can I ask how long do you encode your video, for example your video is 1 hour 30 minutes to 2 hours? On my video, just 2 minutes but I see they encode so slow, 2 hours but just 4mb and dont finish yet. I use I7-9750h
> 
> About the speed. Which preset do you use? faster runs at around 5fps, fast around 1-2 and medium around 1/2 fps. Depending on your CPU and number of used threads.

Which preset is good for quality ? "Slower" preset ?

**Reply by adamjw24** [reply]
Every preset can produce good quality. Quality is mostly determined by the bitrate, not the preset. Preset is only about how much time you want to spend squeezing that last bit of efficiency.

**Reply by adamjw24** [reply]
medium is a good compromise for VVenC.
